pwck.1
.\" Copyright 1992, John F. Haugh II
.\" All rights reserved.
.\"
.\" Permission is granted to copy and create derivative works for any
.\" non-commercial purpose, provided this copyright notice is preserved
.\" in all copies of source code, or included in human readable form
.\" and conspicuously displayed on all copies of object code or
.\" distribution media.
.\"
.\" This software is provided on an AS-IS basis and the author makes
.\" no warrantee of any kind.
.\"
.\"	@(#)pwck.1	3.2	17:43:54	01 May 1993
.\"
.TH PWCK 1
.SH NAME
pwck \- verify integrity of password files
.SH SYNOPSIS
\fBpwck\fR [ \fB-r\fR ] [ \fIpasswd\fR \fIshadow\fR ]
.SH DESCRIPTION
\fBpwck\fR verifies the integrity of the system authentication information.
All entries in the \fB/etc/passwd\fR and \fB/etc/shadow\fR are checked to
see that the entry has the proper format and valid data in each field.
The user is prompted to delete entries that are improperly formatted or
which have other incorrectable errors.
.P
Checks are made to verify that each entry has
.sp
.in +.5i
- the correct number of fields
.br
- a unique user name
.br
- a valid user and group identifier
.br
- a valid primary group
.br
- a valid home directory
.br
- a valid login shell
.in -.5i
.sp
.P
The checks for correct number of fields and unique user name are fatal.
If the entry has the wrong number of fields, the user will be prompted to
delete the entire line.
If the user does not answer affirmatively, all further checks are bypassed.
An entry with a duplicated user name is prompted for deletion, but the
remaining checks will still be made.
All other errors are warning and the user is encouraged to run the
\fBusermod\fR command to correct the error.
.P
The commands which operate on the \fB/etc/passwd\fR file are not able to
alter corrupted or duplicated entries.
\fBpwck\fR should be used in those circumstances to remove the offending
entry.
.SH OPTIONS
By default, \fBpwck\fR operates on the files \fB/etc/passwd\fR and
\fB/etc/shadow\fR.
The user may select alternate files with the \fIpasswd\fR and \fIshadow\fR
parameters.
Additionally, the user may execute the command in read-only mode by
specifying the \fB-r\fR flag.
This causes all questions regarding changes to be answered \fBno\fR
without user intervention.
.SH FILES
/etc/passwd \- user account information
.br
/etc/shadow \- encrypted password information
.br
/etc/group \- group information
.SH SEE ALSO
usermod(1), group(4), passwd(4), shadow(4)
.SH DIAGNOSTICS
The \fIpwck\fR command exits with the following values:
.IP 0 5
Success
.IP 1 5
Syntax Error
.IP 2 5
One or more bad password entries
.IP 3 5
Cannot open password files
.IP 4 5
Cannot lock password files
.IP 5 5
Cannot update password files


NEWS
$Id$

shadow-4.1.5.1 -> shadow-4.2					UNRELEASED

*** general
  * Handle libc whose crypt() returns NULL when passed a salt that
    violates specs or system requirements (e.g. FIPS140). This is needed
    with glibc/eglibc 2.17 for tools checking passwords (passwd (non PAM
    enabled) or newgrp), and for tools generating encrypted passwords
    (chgpasswd, chpasswd, or gpasswd when non PAM enabled or when a fixed
    crypt method is requested on the command line, and newusers, or passwd
    in their non PAM enabled versions)
  * Fix segfault when reading groups split on multiple lines. This impacts
    most user/group management tools when MAX_MEMBERS_PER_GROUP is set.

- su
  * When su receives a signal (SIGTERM, or SIGINT/SIGQUIT in non
    interactive mode), kill the child process group, rather than just the
    immediate child.
  * Fix segmentation faults for users without a proper home or shell in
    their passwd entries.

- login
  * Fix segmentation faults for users without a proper home or shell in
    their passwd entries.

*** documentation
  * Fixed useradd man page (--home-dir option, instead of --home).

*** translation
  * Updated Russian translation.
  * Updated German man pages translation.
  * Fixed gshadow Japanese man page translation.

shadow-4.1.5 -> shadow-4.1.5.1					2012-05-25

- login
  * Log into utmp(x) when PAM is enabled, but do not log into wtmp.
    This complete pam_lastlog which logs into wtmp and in into utmp(x).
- su
  * non PAM enabled versions: do not fail if su is called without a
    controlling terminal.
- userdel
  * Fix segfault when userdel removes the user's group.

*** documentation
  * .so links now point to paths relative to the top-level manual hierarchy

*** translation
  * Updated French man pages translation.
  * Updated German man pages translation.
  * Updated Polish man pages translation. (logoutd.8)

shadow-4.1.4.3 -> shadow-4.1.5					2012-02-12

*** security
  * su -c could be abused by the executed command to invoke commands with
    the caller privileges. See below. (CVE-2005-4890)

*** general
  * report usage error to stderr, but report usage help to stdout (and return
    zero) when explicitly requested (e.g. with --help).
  * initial support for tcb (http://openwall.com/tcb/) for useradd,
    userdel, usermod, chage, pwck, vipw.
  * Added support for ACLs and Extended Attributes in useradd and usermod.
    Support shall be enabled with the new --with-acl or --with-attr
    configure options.
  * Added diagnosis for lock failures.
  * use libsemanage instead of the semanage tool.

- chage
  * Add --root option.
- chfn
  * Add --root option.
- chgpasswd
  * When the gshadow file exists but there are no gshadow entries, an entry
    is created if the password is changed and group requires a
    shadow entry.
  * Add --root option.
- chpasswd
  * PAM enabled versions: restore the -e option to allow restoring
    passwords without knowing those passwords. Restore together the -m
    and -c options. (These options were removed in shadow-4.1.4 on PAM
    enabled versions)
  * When the shadow file exists but there are no shadow entries, an entry
    is created if the password is changed and passwd requires a
    shadow entry.
  * Add --root option.
- chsh
  * Add --root option.
- faillog
  * The -l, -m, -r, -t options only act on the existing users, unless -a is
    specified.
  * Add --root option.
- gpasswd
  * Add --root option.
- groupadd
  * Add --root option.
- groupdel
  * Add --root option.
- groupmems
  * Fix parsing of gshadow entries.
  * Add --root option.
- groupmod
  * Fixed groupmod when configured with --enable-account-tools-setuid.
  * When the gshadow file exists but there are no gshadow entries, an entry
    is created if the password is changed and group requires a
    shadow entry.
  * Add --root option.
- grpck
  * Add --root option.
  * NIS entries were dropped by -s (sort).
- grpconv
  * Add --root option.
- grpunconv
  * Add --root option.
- lastlog
  * Add --root option.
- login
  * Fixed limits support (non PAM enabled versions only)
  * Added support for infinite limits and group based limits (non PAM
    enabled versions only)
  * Fixed infinite loop when CONSOLE is configured with a colon-separated
    list of TTYs.
  * Fixed warning and support for CONSOLE_GROUPS for users member of more
    than 16 groups.
  * Do not log into utmp(x) or wtmp when PAM is enabled. This is done by
    pam_lastlog.
- newgrp, sg
  * Fix parsing of gshadow entries.
- newusers
  * Add --root option.
- passwd
  * Add --root option.
- pwpck
  * NIS entries were dropped by -s (sort).
  * Add --root option.
- pwconv
  * Add --root option.
- pwunconv
  * Add --root option.
- useradd
  * If the skeleton directory contained hardlinked files, copies of the
    hardlink were removed from the skeleton directory.
  * Add --root option.
- userdel
  * Check the existence of the user's mail spool before trying to remove
    it. If it does not exist, a warning is issued, but no failure.
  * Do not remove a group with the same name as the user (usergroup) if
    this group isn't the user's primary group.
  * Add --root option.
  * Add --selinux-user option.
- usermod
  * Accept options in any order (username not necessarily at the end)
  * When the shadow file exists but there are no shadow entries, an entry
    is created if the password is changed and passwd requires a
    shadow entry, or if aging features are used (-e or -f).
  * Add --root option.
- su
  * Document the su exit values.
  * When su receives a signal, wait for the child to terminate (after
    sending a SIGTERM), and kill it only if it did not terminate by itself.
    No delay will be enforced if the child cooperates.
  * Default ENV_SUPATH is /sbin:/bin:/usr/sbin:/usr/bin
  * Fixed infinite loop when CONSOLE is configured with a colon-separated
    list of TTYs.
  * Fixed warning and support for CONSOLE_GROUPS for users member of more
    than 16 groups.

pwck.c
/*
 * Copyright 1992, 1993, John F. Haugh II
 * All rights reserved.
 *
 * Permission is granted to copy and create derivative works for any
 * non-commercial purpose, provided this copyright notice is preserved
 * in all copies of source code, or included in human readable form
 * and conspicuously displayed on all copies of object code or
 * distribution media.
 *
 * This software is provided on an AS-IS basis and the author makes
 * no warrantee of any kind.
 */

#ifndef	lint
static	char	sccsid[] = "@(#)pwck.c	3.2	17:42:44	01 May 1993";
#endif

#include <stdio.h>
#include <fcntl.h>
#include <grp.h>
#ifdef	BSD
#include <strings.h>
#else
#include <string.h>
#endif

#include "config.h"
#include "pwd.h"
#ifdef	SHADOWPWD
#include "shadow.h"
#endif

#ifdef	USE_SYSLOG
#include <syslog.h>

#ifndef	LOG_WARN
#define	LOG_WARN LOG_WARNING
#endif
#endif

struct	pw_file_entry {
	char	*pwf_line;
	int	pwf_changed;
	struct	passwd	*pwf_entry;
	struct	pw_file_entry *pwf_next;
};

#ifdef	SHADOWPWD
struct	spw_file_entry {
	char	*spwf_line;
	int	spwf_changed;
	struct	spwd	*spwf_entry;
	struct	spw_file_entry *spwf_next;
};
#endif

/*
 * Exit codes
 */

#define	E_OKAY		0
#define	E_USAGE		1
#define	E_BADENTRY	2
#define	E_CANTOPEN	3
#define	E_CANTLOCK	4
#define	E_CANTUPDATE	5

/*
 * Message strings
 */

char	*CANTOPEN = "%s: cannot open file %s\n";
char	*CANTLOCK = "%s: cannot lock file %s\n";
char	*CANTUPDATE = "%s: cannot update file %s\n";
char	*CHANGES = "%s: the files have been updated; run mkpasswd\n";
char	*NOCHANGES = "%s: no changes\n";
char	*NOGROUP = "user %s: no group %d\n";
char	*NOHOME = "user %s: directory %s does not exist\n";
char	*NOSHELL = "user %s: program %s does not exist\n";
char	*BADENTRY = "invalid password file entry\n";
char	*PWDUP = "duplicate password entry\n";
char	*DELETE = "delete line `%s'? ";
char	*NO = "No";
#ifdef	SHADOWPWD
char	*BADSENTRY = "invalid shadow password file entry\n";
char	*SPWDUP = "duplicate shadow password entry\n";
#endif

/*
 * Global variables
 */

extern	int	optind;
extern	char	*optarg;
extern	struct	pw_file_entry	*__pwf_head;
extern	int	__pw_changed;
#ifdef	SHADOWPWD
extern	struct	spw_file_entry	*__spwf_head;
extern	int	__sp_changed;
#endif

/*
 * Local variables
 */

char	*Prog;
char	*pwd_file = PWDFILE;
#ifdef	SHADOWPWD
char	*spw_file = SHADOW;
#endif
char	read_only;

/*
 * usage - print syntax message and exit
 */

usage ()
{
#ifdef	SHADOWPWD
	fprintf (stderr, "Usage: %s [ -r ] [ passwd shadow ]\n", Prog);
#else
	fprintf (stderr, "Usage: %s [ -r ] [ passwd ]\n", Prog);
#endif
	exit (E_USAGE);
}

/*
 * yes_or_no - get answer to question from the user
 */

int
yes_or_no ()
{
	char	buf[BUFSIZ];

	/*
	 * In read-only mode all questions are answered "no".
	 */

	if (read_only) {
		puts (NO);
		return 0;
	}

	/*
	 * Get a line and see what the first character is.
	 */

	if (fgets (buf, BUFSIZ, stdin))
		return buf[0] == 'y' || buf[0] == 'Y';

	return 0;
}

/*
 * pwck - verify password file integrity
 */

main (argc, argv)
int	argc;
char	**argv;
{
	int	arg;
	int	errors = 0;
	int	deleted = 0;
	struct	pw_file_entry	*pfe, *tpfe;
	struct	passwd	*pwd;
#ifdef	SHADOWPWD
	struct	spw_file_entry	*spe, *tspe;
	struct	spwd	*spw;
#endif

	/*
	 * Get my name so that I can use it to report errors.
	 */

	if (Prog = strrchr (argv[0], '/'))
		Prog++;
	else
		Prog = argv[0];

#ifdef	USE_SYSLOG
	openlog (Prog, LOG_PID|LOG_CONS|LOG_NOWAIT, LOG_AUTH);
#endif

	/*
	 * Parse the command line arguments
	 */

	while ((arg = getopt (argc, argv, "r")) != EOF) {
		if (arg == 'r')
			read_only = 1;
		else if (arg != EOF)
			usage ();
	}

	/*
	 * Make certain we have the right number of arguments
	 */

#ifdef	SHADOWPWD
	if (optind != argc && optind + 2 != argc)
#else
	if (optind != argc && optind + 1 != argc)
#endif
		usage ();

	/*
	 * If there are two left over filenames, use those as the
	 * password and shadow password filenames.
	 */

	if (optind != argc) {
		pwd_file = argv[optind];
		pw_name (pwd_file);
#ifdef	SHADOWPWD
		spw_file = argv[optind + 1];
		spw_name (spw_file);
#endif
	}

	/*
	 * Lock the files if we aren't in "read-only" mode
	 */

	if (! read_only) {
		if (! pw_lock ()) {
			fprintf (stderr, CANTLOCK, Prog, pwd_file);
#ifdef	USE_SYSLOG
			if (optind == argc)
				syslog (LOG_WARN, "cannot lock %s\n", pwd_file);

			closelog ();
#endif
			exit (E_CANTLOCK);
		}
#ifdef	SHADOWPWD
		if (! spw_lock ()) {
			fprintf (stderr, CANTLOCK, Prog, spw_file);
#ifdef	USE_SYSLOG
			if (optind == argc)
				syslog (LOG_WARN, "cannot lock %s\n", spw_file);

			closelog ();
#endif
			exit (E_CANTLOCK);
		}
#endif
	}

	/*
	 * Open the files.  Use O_RDONLY if we are in read_only mode,
	 * O_RDWR otherwise.
	 */

	if (! pw_open (read_only ? O_RDONLY:O_RDWR)) {
		fprintf (stderr, CANTOPEN, Prog, pwd_file);
#ifdef	USE_SYSLOG
		if (optind == argc)
			syslog (LOG_WARN, "cannot open %s\n", pwd_file);

		closelog ();
#endif
		exit (E_CANTOPEN);
	}
#ifdef	SHADOWPWD
	if (! spw_open (read_only ? O_RDONLY:O_RDWR)) {
		fprintf (stderr, CANTOPEN, Prog, spw_file);
#ifdef	USE_SYSLOG
		if (optind == argc)
			syslog (LOG_WARN, "cannot open %s\n", spw_file);

		closelog ();
#endif
		exit (E_CANTOPEN);
	}
#endif

	/*
	 * Loop through the entire password file.
	 */

	for (pfe = __pwf_head;pfe;pfe = pfe->pwf_next) {

		/*
		 * Start with the entries that are completely corrupt.
		 * They have no (struct passwd) entry because they couldn't
		 * be parsed properly.
		 */

		if (pfe->pwf_entry == (struct passwd *) 0) {

			/*
			 * Tell the user this entire line is bogus and
			 * ask them to delete it.
			 */

			printf (BADENTRY);
			printf (DELETE, pfe->pwf_line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */

			if (! yes_or_no ())
				continue;

			/*
			 * All password file deletions wind up here.  This
			 * code removes the current entry from the linked
			 * list.  When done, it skips back to the top of
			 * the loop to try out the next list element.
			 */

delete_pw:
#ifdef	USE_SYSLOG
			syslog (LOG_INFO,
				"delete passwd line `%s'\n", pfe->pwf_line);
#endif
			deleted++;
			__pw_changed = 1;

			/*
			 * Simple case - delete from the head of the
			 * list.
			 */

			if (pfe == __pwf_head) {
				__pwf_head = pfe->pwf_next;
				continue;
			}

			/*
			 * Hard case - find entry where pwf_next is
			 * the current entry.
			 */

			for (tpfe = __pwf_head;tpfe->pwf_next != pfe;
					tpfe = pfe->pwf_next)
				;

			tpfe->pwf_next = pfe->pwf_next;
			continue;
		}

		/*
		 * Password structure is good, start using it.
		 */

		pwd = pfe->pwf_entry;

		/*
		 * Make sure this entry has a unique name.
		 */

		for (tpfe = __pwf_head;tpfe;tpfe = tpfe->pwf_next) {

			/*
			 * Don't check this entry
			 */

			if (tpfe == pfe)
				continue;

			/*
			 * Don't check invalid entries.
			 */

			if (tpfe->pwf_entry == (struct passwd *) 0)
				continue;

			if (strcmp (pwd->pw_name, tpfe->pwf_entry->pw_name))
				continue;

			/*
			 * Tell the user this entry is a duplicate of
			 * another and ask them to delete it.
			 */

			puts (PWDUP);
			printf (DELETE, pfe->pwf_line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */

			if (yes_or_no ())
				goto delete_pw;
		}

		/*
		 * Make sure the primary group exists
		 */

		if (! getgrgid (pwd->pw_gid)) {

			/*
			 * No primary group, just give a warning
			 */

			printf (NOGROUP, pwd->pw_name, pwd->pw_gid);
			errors++;
		}

		/*
		 * Make sure the home directory exists
		 */

		if (access (pwd->pw_dir, 0)) {

			/*
			 * Home directory doesn't exist, give a warning
			 */

			printf (NOHOME, pwd->pw_name, pwd->pw_dir);
			errors++;
		}

		/*
		 * Make sure the login shell is executable
		 */

		if (pwd->pw_shell[0] && access (pwd->pw_shell, 0)) {

			/*
			 * Login shell doesn't exist, give a warning
			 */
			
			printf (NOSHELL, pwd->pw_name, pwd->pw_shell);
			errors++;
		}
	}

#ifdef	SHADOWPWD
	/*
	 * Loop through the entire shadow password file.
	 */

	for (spe = __spwf_head;spe;spe = spe->spwf_next) {

		/*
		 * Start with the entries that are completely corrupt.
		 * They have no (struct spwd) entry because they couldn't
		 * be parsed properly.
		 */

		if (spe->spwf_entry == (struct spwd *) 0) {

			/*
			 * Tell the user this entire line is bogus and
			 * ask them to delete it.
			 */

			printf (BADSENTRY);
			printf (DELETE, spe->spwf_line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */

			if (! yes_or_no ())
				continue;

			/*
			 * All shadow file deletions wind up here.  This
			 * code removes the current entry from the linked
			 * list.  When done, it skips back to the top of
			 * the loop to try out the next list element.
			 */

delete_spw:
#ifdef	USE_SYSLOG
			syslog (LOG_INFO,
				"delete shadow line `%s'\n", spe->spwf_line);
#endif
			deleted++;
			__sp_changed = 1;

			/*
			 * Simple case - delete from the head of the
			 * list.
			 */

			if (spe == __spwf_head) {
				__spwf_head = spe->spwf_next;
				continue;
			}

			/*
			 * Hard case - find entry where spwf_next is
			 * the current entry.
			 */

			for (tspe = __spwf_head;tspe->spwf_next != spe;
					tspe = spe->spwf_next)
				;

			tspe->spwf_next = spe->spwf_next;
			continue;
		}

		/*
		 * Shadow password structure is good, start using it.
		 */

		spw = spe->spwf_entry;

		/*
		 * Make sure this entry has a unique name.
		 */

		for (tspe = __spwf_head;tspe;tspe = tspe->spwf_next) {

			/*
			 * Don't check this entry
			 */

			if (tspe == spe)
				continue;

			/*
			 * Don't check invalid entries.
			 */

			if (tspe->spwf_entry == (struct spwd *) 0)
				continue;

			if (strcmp (spw->sp_namp, tspe->spwf_entry->sp_namp))
				continue;

			/*
			 * Tell the user this entry is a duplicate of
			 * another and ask them to delete it.
			 */

			puts (SPWDUP);
			printf (DELETE, spe->spwf_line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */

			if (yes_or_no ())
				goto delete_spw;
		}
	}
#endif

	/*
	 * All done.  If there were no deletions we can just abandon any
	 * changes to the files.
	 */

	if (deleted) {
		if (! pw_close ()) {
			fprintf (stderr, CANTUPDATE, Prog, pwd_file);
#ifdef	USE_SYLOG
			syslog (LOG_WARN, "cannot update %s\n", pwd_file);
			closelog ();
#endif
			exit (E_CANTUPDATE);
		}
#ifdef	SHADOWPWD
		if (! spw_close ()) {
			fprintf (stderr, CANTUPDATE, Prog, spw_file);
#ifdef	USE_SYLOG
			syslog (LOG_WARN, "cannot update %s\n", spw_file);
			closelog ();
#endif
			exit (E_CANTUPDATE);
		}
#endif
	}

	/*
	 * Don't be anti-social - unlock the files when you're done.
	 */

#ifdef	SHADOWPWD
	(void) spw_unlock ();
#endif
	(void) pw_unlock ();

	/*
	 * Tell the user what we did and exit.
	 */

	if (errors)
		printf (deleted ? CHANGES:NOCHANGES, Prog);

#ifdef	USE_SYSLOG
	closelog ();
#endif
	exit (errors ? E_BADENTRY:E_OKAY);
}
